[POS][Custom payment UI] – Payment processing animation with Lottie

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
### Animated payment states with Lottie

Closes: #13068
Closes: #12826
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds animations for payment states:
* Collecting
* Capturing
* Processing

See the video below.

Decision on introducing Lottie: p1733860533372109-slack-C070SJRA8DP

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
1. Go to Woo POS (Menu more > Point of Sale)
2. Add items to cart and click "Checkout"
3. Proceed with payment flow
4. Observe animated "card" icon 

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
- [ ] Animation should be visible in Totals pane when the reader is ready (payment Collecting state)
- [ ] Animation should be displayed in fullscreen in processing and capturing states

Design specs: TfaZ4LUkEwEGrxfnEFzvJj-fi-2710_118241
Animation specs: p6riRB-aNq-p2

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
I tested the payment collection in POS and tried tuning up the animation to match specs as much as possible in reasonable time constraints.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/f01f3ee1-5bb4-4f78-b20b-bd465ea427fb

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Show subtitles for `processing` and `capturing` states

samiuelson committed:
Show lottie animations for payment capturing and processing

samiuelson committed:
Revert accidental gradle.properties changes

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Animate "reader ready" state in Totals

samiuelson committed:
Revert accidental change to gradle.properties

samiuelson committed:
Revert accidental change to gradle.properties

samiuelson committed:
Update unit tests

samiuelson committed:
Add tests:

* given order draft created and reader connected, when payment is processed, should show processing state

* given order draft created and reader connected, when payment is captured, should show processing state

samiuelson committed:
Add payment processing screen enter animation

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Tune up animation

samiuelson committed:
Clean up code

samiuelson committed:
Clean up tests

samiuelson committed:
Remove redundant code

samiuelson committed:
Remove redundant code

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:
```diff
@@ -260,6 +256,8 @@ class WooPosTotalsViewModel @Inject constructor(
                         childrenToParentEventSender.sendToParent(ChildToParentEvent.PaymentFailed)
                     }
 
+                    CardReaderPaymentState.ReFetchingOrder -> Unit
```
💡 `ReFetchingOrder` state is emitted after `PaymentSuccess` state on which we show Payment success screen. `ReFetchingOrder` indicates the order is synced in DB. We want to ignore it in POS.

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsScreen.kt:
```diff
@@ -176,11 +182,12 @@ private fun PreparingReader(readerStatus: WooPosTotalsViewState.ReaderStatus) {
+    LottieAnimation(
+        modifier = Modifier.size(256.dp),
+        composition = tapCardAnimation,
+        clipSpec = LottieClipSpec.Markers("reader_awaiting_start", "reader_awaiting_end"),
```
💡 Marker names refer to makers[] table from `woopos_card_ilustration.json` file and point to animation frames we loop animation between.

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Merge branch `custom-payment-ui-6-use-payment-controller-in-pos`

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:
```diff
@@ -227,15 +227,13 @@ class WooPosTotalsViewModel @Inject constructor(
-                    CardReaderPaymentState.ReFetchingOrder -> {
-                        uiState.value = buildPaymentProcessingState()
+                    is CardReaderPaymentState.PaymentCapturing -> {
+                        uiState.value = buildPaymentProcessingState(paymentState)
```
🔍 I'd personally consider either renaming `processing` to something else or splitting processing and capturing steps.

The current solution is fine, but it can be a bit misleading. We could perhaps call it `buildPaymentInProgressState` and `PaymentInProgress` or something in that sense.

samiuelson replied:
Good point. I got inspired by designs in Figma. Renamed in fadb13d9b62e4bc98243638d7820024c11edd62e.

malinajirka commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModelTest.kt:
```diff
@@ -897,6 +866,72 @@ class WooPosTotalsViewModelTest {
         assertThat(totalState.readerStatus).isInstanceOf(WooPosTotalsViewState.ReaderStatus.ReadyForPayment::class.java)
     }
 
+    @Test
```
💬  This is not an issue with this PR since this PR is consistent with the remaining unit tests. However, I think it's something we should iterate on rather sooner than later. IMO the body of each tests should contain only changes that are affecting the result of the test. 

For example, for test called `given order draft created and reader connected, when payment is captured, should show processing state`, the body of the test would IMO ideally prepare only `order draft, reader connected, payment capture`. Everything else would be setup using a single preparation method or ideally in `@Before` or similar. The current approach increases maintenance cost as well as makes it more time consuming to write additional tests.

If implementing such approach is difficult, it's in my experience usually a flag that the SUT is doing more than it should and should be split into smaller classes.

Wdyt?

samiuelson replied:
I agree. `WooPosTotalsViewModelTest` got quite messy recently. I'll try to clean it up as much as possible in next PR and report an issue to team backlog if needed.

malinajirka replied:
Noice, thanks so much! This should definitely be a team effort considering the tests were written by all of us, so don't hesitate spliting it into tasks.

malinajirka reviewed (approved):
Thanks @samiuelson ! Looks great overall. I've left two unrelated comments that don't block merge of this PR.

The animations of the illustrations look great. We'll need to work on both the state transition (transition from one illustration to the other) and screen transitions. These both look significantly smoother on iOS. However, I guess these aren't part of this PR.

I'm leaving the merge to you since I'm not sure if this is dependant on another PR or can be merged immediatelly into `[custom-payment-ui-6-use-payment-controller-in-pos](https://github.com/woocommerce/woocommerce-android/tree/custom-payment-ui-6-use-payment-controller-in-pos)`.

malinajirka reviewed (changes requested):
I encountered this crash when adding a variation into the cart. I thought it was not related to this PR, so I reported it on Slack. However, it seems it works as expected on `trunk`. 

1. Open POS
2. Tap on a variable product
3. Tap on a variation
4. Tap on checkout
5. Notice crash

`
FATAL EXCEPTION: main
                                                                                                    Process: com.woocommerce.android.prealpha, PID: 19229
                                                                                                    java.lang.NullPointerException
                                                                                                    	at com.woocommerce.android.ui.woopos.home.totals.WooPosTotalsRepository$createOrderWithProducts$3$2.invokeSuspend(WooPosTotalsRepository.kt:47)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:111)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:99)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:811)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:715)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:702)
                                                                                                    	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@686a376, Dispatchers.Main.immediate]
`

P.S. Maybe this branch is just out-of-date with Trunk and that's why it crashes?

samiuelson committed:
Merge branch `custom-payment-ui-6-use-payment-controller-in-pos`

samiuelson committed:
Rename Processing UI state to InProgress

samiuelson commented:
> I encountered this crash when adding a variation into the cart. I thought it was not related to this PR, so I reported it on Slack. However, it seems it works as expected on `trunk`.
> 
> 1. Open POS
> 2. Tap on a variable product
> 3. Tap on a variation
> 4. Tap on checkout
> 5. Notice crash

Thanks for the review, @malinajirka. I've updated the branch with the latest trunk and verified the crash is no longer happening. I also addressed comments. Let me know what you think!

samiuelson committed:
Merge branch 'custom-payment-ui-6-use-payment-controller-in-pos' into custom-payment-ui-payment-processing-animation

malinajirka reviewed (approved):
Awesome, thanks for the follow ups! LGTM 🙇
